File: mobile/android/core/designsystem/src/main/kotlin/com/slopssaloon/omen/core/designsystem/component/OmenContextualHelp.kt
package com.slopssaloon.omen.core.designsystem.component

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.semantics
import com.slopssaloon.omen.core.designsystem.theme.OmenTheme

/**
 * Registry §3.1 **Tooltip / Help**, built for M6-ContextualHelp.
 *
 * Authority: `Blueprints/specs/mobile/m4-help-support-v1.md` §1–§5. This is the contextual half
 * of that contract — a short explanation of the surface a person is already looking at. The
 * durable Help + Support destination is `OmenHelpSupportScreen` and is untouched here.
 *
 * Rules this component exists to enforce (spec §2):
 * - never unsolicited — nothing but the person's own tap can open it;
 * - never blocks a decision, so it owns no confirm/deny action and gates nothing;
 * - dismissing returns to the exact prior state, which is why it holds no state beyond its own
 *   visibility and reports nothing back to its host.
 *
 * The DS module stays product-agnostic: the types below carry no Omen concept, and the
 * per-destination copy lives in the feature layer at `app/feature/help/ContextualHelpContent.kt`.
 */
data class OmenHelpTip(val label: String, val body: String)

/**
 * A short, local explanation of one surface.
 *
 * Spec §4: "A contextual surface that needs more than a short explanation routes to the durable
 * Help Center instead of becoming a dense tooltip." [tips] is capped at [MAX_TIPS] so that stays
 * a rule rather than a matter of taste.
 */
data class OmenHelpTopic(
    val title: String,
    val summary: String,
    val tips: List<OmenHelpTip>,
) {
    companion object {
        /** Above this, content belongs in Help + Support, not in a contextual surface. */
        const val MAX_TIPS = 4
    }
}

/**
 * The `What is this?` affordance: an icon-only control that presents its topic on tap.
 *
 * Built on [OmenIconButton] so the 48dp touch target, focus ring, and required content
 * description come from the approved primitive rather than being restated here.
 *
 * [icon] is a slot rather than a built-in glyph: drawable resources live in the app module
 * (`res/drawable/ic_help.xml`), and `:core:designsystem` stays free of product resources. The
 * feature-layer `OmenHelpButton` supplies it, so no screen wires the glyph by hand.
 */
@Composable
fun OmenContextualHelpButton(
    topic: OmenHelpTopic,
    modifier: Modifier = Modifier,
    size: OmenIconButtonSize = OmenIconButtonSize.Md,
    icon: @Composable () -> Unit,
) {
    // rememberSaveable: a rotation while help is open should not silently close it.
    var visible by rememberSaveable(topic.title) { mutableStateOf(false) }

    OmenIconButton(
        // TalkBack reads the purpose *and* what will be explained, so the control stays
        // distinguishable when several sit on one screen (spec §5).
        contentDescription = "What is this? ${topic.title}",
        onClick = { visible = true },
        modifier = modifier,
        tone = OmenIconButtonTone.Neutral,
        size = size,
        icon = icon,
    )

    OmenContextualHelpSheet(
        topic = topic,
        visible = visible,
        // System Back dismisses this sheet before leaving the originating screen (spec §4),
        // which ModalBottomSheet routes through onDismissRequest for us.
        onDismissRequest = { visible = false },
    )
}

/** The presented explanation. Reading order is title, summary, then tips (spec §5). */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun OmenContextualHelpSheet(
    topic: OmenHelpTopic,
    visible: Boolean,
    onDismissRequest: () -> Unit,
    modifier: Modifier = Modifier,
) {
    if (!visible) return
    val sheetState = rememberModalBottomSheetState()
    val spacing = OmenTheme.spacing
    val colors = OmenTheme.color

    ModalBottomSheet(
        onDismissRequest = onDismissRequest,
        sheetState = sheetState,
        modifier = modifier,
        containerColor = colors.surface2,
        contentColor = colors.textPrimary,
    ) {
        Column(
            Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                // At large font scales the tips outrun the sheet's default height. The
                // content scrolls, and this keeps the last line clear of the gesture bar
                // instead of ending underneath it.
                .navigationBarsPadding()
                .padding(horizontal = spacing.cardInterior)
                .padding(bottom = spacing.bodyToFooter),
            verticalArrangement = Arrangement.spacedBy(spacing.headerToBody),
        ) {
            Text(
                topic.title,
                style = OmenTheme.typography.h2.toTextStyle(),
                color = colors.textPrimary,
                modifier = Modifier.semantics { heading() },
            )
            Text(
                topic.summary,
                style = OmenTheme.typography.body.toTextStyle(),
                // Registry §3.1 Tooltip/Help names `surface-2` + `text-primary`. That pairing
                // is not decorative: `text-secondary` on `surface-2` measures 4.43:1 in light
                // mode — under AA. Caught on iOS by Apple's accessibility audit
                // ("Contrast nearly passed") and corrected on both platforms.
                color = colors.textPrimary,
            )
            Column(verticalArrangement = Arrangement.spacedBy(spacing.fieldToField)) {
                topic.tips.forEach { tip ->
                    Column(verticalArrangement = Arrangement.spacedBy(spacing.inputToHint)) {
                        Text(
                            tip.label,
                            style = OmenTheme.typography.label.toTextStyle(),
                            color = colors.accent,
                        )
                        Text(
                            tip.body,
                            style = OmenTheme.typography.bodySmall.toTextStyle(),
                            color = colors.textPrimary,
                        )
                    }
                }
            }
        }
    }
}

File: mobile/ios/OmenIOS/OmenIOS/DesignSystem/OmenContextualHelp.swift
import SwiftUI

/// Registry §3.1 **Tooltip / Help**, built for M6-ContextualHelp.
///
/// Authority: `Blueprints/specs/mobile/m4-help-support-v1.md` §1–§5. This is the first half of
/// that contract — "explains the current Omen concept, state, or next step without taking the
/// person away from their work". The second half, the durable Help + Support destination, is
/// `OmenHelpSupportView` and is deliberately untouched here.
///
/// Rules this type exists to enforce (spec §2):
/// - it is **never** unsolicited — presentation is only ever driven by the person's own tap;
/// - it never blocks a decision, so it owns no confirm/deny action and gates nothing;
/// - dismissing returns to the exact prior state, which is why it holds no state but its own
///   `isPresented` and reports nothing back to its host.
///
/// The DS module stays product-agnostic (registry §3.2 M4 note): the value types below carry
/// no Omen concept, and the per-destination copy lives in the feature layer at
/// `App/Help/OmenContextualHelpContent.swift`.

/// One labeled explanation inside a help topic.
struct OmenHelpTip: Equatable, Identifiable {
    let label: String
    let body: String

    var id: String { label }

    init(label: String, body: String) {
        self.label = label
        self.body = body
    }
}

/// A short, local explanation of one surface.
///
/// Spec §4: "A contextual surface that needs more than a short explanation routes to the
/// durable Help Center instead of becoming a dense tooltip." `tips` is capped at
/// ``OmenHelpTopic/maxTips`` to keep that from being a matter of taste.
struct OmenHelpTopic: Equatable {
    let title: String
    let summary: String
    let tips: [OmenHelpTip]

    /// Above this, content belongs in Help + Support, not in a contextual surface.
    static let maxTips = 4

    init(title: String, summary: String, tips: [OmenHelpTip]) {
        self.title = title
        self.summary = summary
        self.tips = tips
    }
}

/// The `What is this?` affordance: an icon-only control that presents its topic on tap.
///
/// Renders as an `OmenIconButton` rather than a bespoke control so the 44pt touch target,
/// focus ring, and required accessibility name come from the approved primitive.
struct OmenContextualHelpButton: View {
    let topic: OmenHelpTopic
    var size: OmenIconButtonSize = .md

    @State private var isPresented = false

    var body: some View {
        OmenIconButton(
            // VoiceOver reads the purpose *and* what will be explained, so the control is
            // distinguishable when several sit on one screen (spec §5).
            contentDescription: "What is this? \(topic.title)",
            icon: Image(systemName: "questionmark.circle"),
            action: { isPresented = true },
            tone: .neutral,
            size: size
        )
        .sheet(isPresented: $isPresented) {
            OmenContextualHelpSheet(topic: topic, onDismiss: { isPresented = false })
                // Short by contract, so the sheet opens at a height that leaves the
                // originating screen visible behind it — help beside the work, not over it.
                .presentationDetents([.medium, .large])
                .presentationDragIndicator(.visible)
        }
    }
}

/// The presented explanation. Reading order is title, summary, then tips (spec §5).
struct OmenContextualHelpSheet: View {
    let topic: OmenHelpTopic
    let onDismiss: () -> Void

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: OmenSpacing.headerToBody) {
                HStack(alignment: .firstTextBaseline) {
                    Text(topic.title)
                        .omenTextStyle(OmenTypography.h2)
                        .foregroundStyle(OmenColor.textPrimary)
                        .accessibilityAddTraits(.isHeader)
                    Spacer(minLength: OmenSpacing.step8)
                    OmenButton(title: "Done", action: onDismiss, variant: .link, size: .sm)
                }

                Text(topic.summary)
                    .omenTextStyle(OmenTypography.body)
                    // Registry §3.1 Tooltip/Help names `surface-2` + `text-primary`. That
                    // pairing is not decorative: `text-secondary` on `surface-2` measures
                    // 4.43:1 in light mode — under AA — and Apple's accessibility audit
                    // fails it as "Contrast nearly passed".
                    .foregroundStyle(OmenColor.textPrimary)
                    // Large text reflows rather than truncating (spec §4).
                    .fixedSize(horizontal: false, vertical: true)

                if !topic.tips.isEmpty {
                    VStack(alignment: .leading, spacing: OmenSpacing.fieldToField) {
                        ForEach(topic.tips) { tip in
                            VStack(alignment: .leading, spacing: OmenSpacing.inputToHint) {
                                Text(tip.label)
                                    .omenTextStyle(OmenTypography.label)
                                    // `text-primary`, not `accent` — registry §3.1's Tooltip/Help
                                    // row allows exactly `surface-2` + `text-primary`, and brass
                                    // here was an implementation deviation, not an approved
                                    // variant. It also failed WCAG AA in dark mode: measured
                                    // 3.68:1 for accent `#A67C2E` on surface-2 `#2C2C2E`
                                    // (light mode passed at 5.70:1, so this was dark-only).
                                    // The body below already followed the registry; the label
                                    // was the half of that earlier fix that got left behind.
                                    // Hierarchy still reads: `OmenTypography.label` differs from
                                    // the body's `bodySmall` in size and weight, not just colour.
                                    .foregroundStyle(OmenColor.textPrimary)
                                Text(tip.body)
                                    .omenTextStyle(OmenTypography.bodySmall)
                                    .foregroundStyle(OmenColor.textPrimary)
                                    .fixedSize(horizontal: false, vertical: true)
                            }
                            // One tip = one VoiceOver stop, label then body.
                            .accessibilityElement(children: .combine)
                        }
                    }
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(OmenSpacing.cardInterior)
        }
        .background(OmenColor.surface2)
    }
}
